test/benchmarks.cpp

```diff
@@ -0,0 +1,460 @@
+//////////////////////////////////////////////////////////////////////////////////////
+//
+// (C) Daniel Strano 2017, 2018. All rights reserved.
+//
+// This is a multithreaded, universal quantum register simulation, allowing
+// (nonphysical) register cloning and direct measurement of probability and
+// phase, to leverage what advantages classical emulation of qubits can have.
+//
+// Licensed under the GNU General Public License V3.
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
+// for details.
+
+#include <atomic>
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "catch.hpp"
+#include "qfactory.hpp"
+
+#include "tests.hpp"
+
+using namespace Qrack;
+
+#define EPSILON 0.001
+#define REQUIRE_FLOAT(A, B) do {                                            \
+        double __tmp_a = A;                                                 \
+        double __tmp_b = B;                                                 \
+        REQUIRE(__tmp_a < (__tmp_b + EPSILON));                             \
+        REQUIRE(__tmp_b > (__tmp_b - EPSILON));                             \
+    } while (0);
+
+void benchmarkLoop(std::function<void(QInterfacePtr, int)> fn) {
+
+    const int ITERATIONS = 100;
+
+    std::cout<<std::endl;
+    std::cout<<ITERATIONS<<" iterations";
+    std::cout<<std::endl;
+    std::cout<<"# of Qubits, ";
+    std::cout<<"Average Time (ms), ";
+    std::cout<<"Sample Std. Deviation (ms), ";
+    std::cout<<"Fastest (ms), ";
+    std::cout<<"1st Quartile (ms), ";
+    std::cout<<"Median (ms)";
+    std::cout<<"3rd Quartile (ms), ";
+    std::cout<<"Slowest (ms)"<<std::endl;
+
+    clock_t startClock, tClock, allClock;
+    clock_t trialClocks[ITERATIONS];
+
+    int i, numBits;
+
+    double avgt, stdet;
+
+    // Grover's search inverts the function of a black box subroutine.
+    // Our subroutine returns true only for an input of 100.
+    for (numBits = 3; numBits <= 20; numBits++) {
+        QInterfacePtr qftReg = CreateQuantumInterface(testEngineType, testSubEngineType, numBits, 0, rng);
+        startClock = clock();
+        allClock = startClock;
+        tClock = startClock;
+        avgt = 0.0;
+        for (i = 0; i < ITERATIONS; i++) {
+
+            // Run loop body
+            fn(qftReg, numBits);
+
+            // Display timing
+            tClock = clock() - allClock;
+            trialClocks[i] = tClock;
+            avgt += tClock;
+            allClock = clock();
```

Comment:
`allClock` is really a misleading name.  Use `iterClock`.

Also, is there a reason why this is set at the bottom of the loop instead (as would be normally expected) at the top?

Reply:
See my comment on the previous review. The initialization is not a special case of the iteration. This would be off-by-one.

Reply:
The point is that the trial timing is calculated immediately at the end of each iteration. We have two options: find a way to make the initialization a case of the loop, and calculate one additional time after the loop is done. Or, make the initialization it's own case, and calculate the timing at the bottom. To just do it in the loop without having either a special first or last case is off-by-one.

Reply:
`allClock` has been changed to `iterClock`.